test/relayer.js

```diff
@@ -0,0 +1,351 @@
+        "RM: Wrong number of signatures"
+      );
+    });
+
+    it("should fail a duplicate transaction", async () => {
+      const nonce = await utils.getNonceForRelay();
+      const chainId = await utils.getChainId();
+      const gasLimit = 100000;
+      const transaction = await encodeTransaction(recipient, 10, ZERO_BYTES32);
+      const methodData = module.contract.methods.multiCall(wallet.address, [transaction]).encodeABI();
+      const signatures = await utils.signOffchain(
+        [owner],
+        module.address,
+        0,
+        methodData,
+        chainId,
+        nonce,
+        0,
+        gasLimit,
+        ETH_TOKEN,
+        ethers.constants.AddressZero,
+      );
+
+      await module.execute(
+        wallet.address,
+        methodData,
+        nonce,
+        signatures,
+        0,
+        gasLimit,
+        ETH_TOKEN,
+        ethers.constants.AddressZero);
+
+      await truffleAssert.reverts(
+        module.execute(
+          wallet.address,
+          methodData,
+          nonce,
+          signatures,
+          0,
+          gasLimit,
+          ETH_TOKEN,
+          ethers.constants.AddressZero),
+        "RM: Duplicate request");
+    });
+
+    it("should update the nonce after the transaction", async () => {
+      const transaction = await encodeTransaction(recipient, 10, ZERO_BYTES32);
+
+      const nonceBefore = await module.getNonce(wallet.address);
+      await manager.relay(
+        module,
+        "multiCall",
+        [wallet.address, [transaction]],
+        wallet,
+        [owner],
+        1,
+        ETH_TOKEN,
+        relayer);
+      const nonceAfter = await module.getNonce(wallet.address);
+      expect(nonceAfter).to.be.gt.BN(nonceBefore);
+    });
+  });
+
+  describe("refund", () => {
+    beforeEach(async () => {
+      await initStorage();
+      await whitelist(recipient);
+    });
+
+    it("should refund in ETH", async () => {
+      // eth balance
+      const balanceStart = await utils.getBalance(wallet.address);
+      // send erc20
+      const data = token.contract.methods.transfer(recipient, 100).encodeABI();
+      const transaction = await encodeTransaction(token.address, 0, data, true);
+
+      const txReceipt = await manager.relay(
+        module,
+        "multiCall",
+        [wallet.address, [transaction]],
+        wallet,
+        [owner],
+        1,
+        ETH_TOKEN,
+        relayer);
+      const success = await utils.parseRelayReceipt(txReceipt).success;
+      assert.isTrue(success, "transfer failed");
+      const balanceEnd = await utils.getBalance(wallet.address);
+      expect(balanceEnd.sub(balanceStart)).to.be.lt.BN(0);
+    });
+
+    it("should refund in ERC20", async () => {
+      // token balance
+      const balanceStart = await token.balanceOf(relayer);
+      // send ETH
+      const transaction = await encodeTransaction(recipient, 10, ZERO_BYTES32, false);
+      const txReceipt = await manager.relay(
```

Comment:
This is failing on `(wethReserve, tokenReserve,) = IUniswapV2Pair(pair).getReserves();` but I haven't investigated further

Reply:
Thanks, that helps and I know what the problem is.